org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/text/correction/ReturnTypeSubProcessor.java

```diff
@@ -13,389 +13,101 @@
-					proposals.add(new ReplaceCorrectionProposal(label, cu, nameNode.getStartPosition(), nameNode.getLength(), constructorName, IProposalRelevance.CHANGE_TO_CONSTRUCTOR));
-				}
-			}
-		}
+	public static void replaceReturnWithYieldStatementProposals(IInvocationContext context, IProblemLocationCore problem, Collection<ICommandAccess> proposals) {
+		new ReturnTypeSubProcessor().collectReplaceReturnWithYieldStatementProposals(context, problem, proposals);
 	}
 
-	public static void addMissingReturnStatementProposals(IInvocationContext context, IProblemLocationCore problem, Collection<ICommandAccess> proposals) {
-		ICompilationUnit cu= context.getCompilationUnit();
+	public static void addMethodRetunsVoidProposals(IInvocationContext context, IProblemLocationCore problem, Collection<ICommandAccess> proposals) throws JavaModelException {
+		new ReturnTypeSubProcessor().collectMethodReturnsVoidProposals(context, problem, proposals);
+	}
 
-		ASTNode selectedNode= problem.getCoveringNode(context.getASTRoot());
-		if (selectedNode == null) {
-			return;
-		}
-		ReturnStatement existingStatement= (selectedNode instanceof ReturnStatement) ? (ReturnStatement) selectedNode : null;
-		// Lambda Expression can be in a MethodDeclaration or a Field Declaration
-		if (selectedNode instanceof LambdaExpression) {
-			MissingReturnTypeInLambdaCorrectionProposal proposal= new MissingReturnTypeInLambdaCorrectionProposal(cu, (LambdaExpression) selectedNode, existingStatement,
-					IProposalRelevance.MISSING_RETURN_TYPE);
-			proposals.add(proposal);
-		} else {
-			BodyDeclaration decl= ASTResolving.findParentBodyDeclaration(selectedNode);
-			if (decl instanceof MethodDeclaration) {
-				MethodDeclaration methodDecl= (MethodDeclaration) decl;
-				Block block= methodDecl.getBody();
-				if (block == null) {
-					return;
-				}
-				proposals.add(new MissingReturnTypeCorrectionProposal(cu, methodDecl, existingStatement, IProposalRelevance.MISSING_RETURN_TYPE));
 
-				Type returnType= methodDecl.getReturnType2();
-				if (returnType != null && !"void".equals(ASTNodes.asString(returnType))) { //$NON-NLS-1$
-					AST ast= methodDecl.getAST();
-					ASTRewrite rewrite= ASTRewrite.create(ast);
-					rewrite.replace(returnType, ast.newPrimitiveType(PrimitiveType.VOID), null);
-					Javadoc javadoc= methodDecl.getJavadoc();
-					if (javadoc != null) {
-						TagElement tagElement= JavadocTagsSubProcessor.findTag(javadoc, TagElement.TAG_RETURN, null);
-						if (tagElement != null) {
-							rewrite.remove(tagElement, null);
-						}
-					}
 
-					String label= CorrectionMessages.ReturnTypeSubProcessor_changetovoid_description;
-					Image image= JavaPluginImages.get(JavaPluginImages.IMG_CORRECTION_CHANGE);
-					ASTRewriteCorrectionProposal proposal= new ASTRewriteCorrectionProposal(label, cu, rewrite, IProposalRelevance.CHANGE_RETURN_TYPE_TO_VOID, image);
-					proposals.add(proposal);
-				}
-			}
+	private static Image findImage(int id) {
```

Comment:
I think that the codes should be constants which correspond to the callers (e.g. MissingReturnTypeProposal1).  If in this case all the corrections will be changes, make it the default instead of null and avoid having to code the switch.